// @microsoft/office-js-helpers/dist/helpers/dialog.js
/* Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license. See LICENSE in the project root for license information. */
"use strict";
var utilities_1 = require("./utilities");
var dialog_1 = require("../errors/dialog");
var Dialog = (function () {
    /**
     * @constructor
     *
     * @param url Url to be opened in the dialog.
     * @param width Width of the dialog.
     * @param height Height of the dialog.
    */
    function Dialog(url, width, height, useTeamsDialog) {
        if (url === void 0) { url = location.origin; }
        if (width === void 0) { width = 1024; }
        if (height === void 0) { height = 768; }
        if (useTeamsDialog === void 0) { useTeamsDialog = false; }
        this.url = url;
        this.width = width;
        this.height = height;
        this.useTeamsDialog = useTeamsDialog;
        if (!(/^https/.test(url))) {
            throw new dialog_1.DialogError('URL has to be loaded over HTTPS.');
        }
        this.size = this._optimizeSize(width, height);
    }
    Object.defineProperty(Dialog.prototype, "result", {
        get: function () {
            if (this._result == null) {
                this._result = this.useTeamsDialog ? this._teamsDialog() : this._addinDialog();
            }
            return this._result;
        },
        enumerable: true,
        configurable: true
    });
    Dialog.prototype._addinDialog = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            Office.context.ui.displayDialogAsync(_this.url, { width: _this.size.width$, height: _this.size.height$ }, function (result) {
                if (result.status === Office.AsyncResultStatus.Failed) {
                    throw new dialog_1.DialogError(result.error.message);
                }
                else {
                    var dialog_2 = result.value;
                    dialog_2.addEventHandler(Office.EventType.DialogMessageReceived, function (args) {
                        try {
                            var result_1 = _this._safeParse(args.message);
                            if (result_1.parse) {
                                resolve(_this._safeParse(result_1.value));
                            }
                            else {
                                resolve(result_1.value);
                            }
                        }
                        catch (exception) {
                            reject(new dialog_1.DialogError('An unexpected error in the dialog has occured.', exception));
                        }
                        finally {
                            dialog_2.close();
                        }
                    });
                    dialog_2.addEventHandler(Office.EventType.DialogEventReceived, function (args) {
                        try {
                            reject(new dialog_1.DialogError(args.message, args.error));
                        }
                        catch (exception) {
                            reject(new dialog_1.DialogError('An unexpected error in the dialog has occured.', exception));
                        }
                        finally {
                            dialog_2.close();
                        }
                    });
                }
            });
        });
    };
    Dialog.prototype._teamsDialog = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            try {
                microsoftTeams.initialize();
            }
            catch (e) {
            }
            microsoftTeams.authentication.authenticate({
                url: _this.url,
                width: _this.size.width,
                height: _this.size.height,
                failureCallback: function (exception) { return reject(new dialog_1.DialogError('Error while launching dialog', exception)); },
                successCallback: function (message) { return resolve(message); }
            });
        });
    };
    /**
     * Close any open dialog by providing an optional message.
     * If more than one dialogs are attempted to be opened
     * an expcetion will be created.
     */
    Dialog.close = function (message, useTeamsDialog) {
        if (useTeamsDialog === void 0) { useTeamsDialog = false; }
        var parse = false;
        var value = message;
        if ((!(value == null)) && typeof value === 'object') {
            parse = true;
            value = JSON.stringify(value);
        }
        else if (typeof message === 'function') {
            throw new dialog_1.DialogError('Invalid message. Canno\'t pass functions as arguments');
        }
        try {
            if (useTeamsDialog) {
                try {
                    microsoftTeams.initialize();
                }
                catch (e) {
                }
                microsoftTeams.authentication.notifySuccess(JSON.stringify({ parse: parse, value: value }));
            }
            else if (utilities_1.Utilities.isAddin) {
                Office.context.ui.messageParent(JSON.stringify({ parse: parse, value: value }));
            }
        }
        catch (error) {
            throw new dialog_1.DialogError('Canno\'t close dialog', error);
        }
    };
    Dialog.prototype._getSize = function (width, height) {
        var screenWidth = window.screen.width;
        if (width && height) {
            return this._optimizeSize(width, height);
        }
        else if (screenWidth <= 640) {
            return this._optimizeSize(640, 480);
        }
        else if (screenWidth <= 1366) {
            return this._optimizeSize(1024, 768);
        }
        else if (screenWidth <= 1920) {
            return this._optimizeSize(1600, 900);
        }
    };
    Dialog.prototype._optimizeSize = function (width, height) {
        var screenWidth = window.screen.width;
        var screenHeight = window.screen.height;
        var optimizedWidth = this._maxSize(width, screenWidth);
        var optimizedHeight = this._maxSize(height, screenHeight);
        return {
            width$: this._percentage(optimizedWidth, screenWidth),
            height$: this._percentage(optimizedHeight, screenHeight),
            width: optimizedWidth,
            height: optimizedHeight
        };
    };
    Dialog.prototype._maxSize = function (value, max) {
        return value < (max - 30) ? value : max - 30;
    };
    ;
    Dialog.prototype._percentage = function (value, max) {
        return (value * 100 / max);
    };
    Dialog.prototype._safeParse = function (data) {
        try {
            var result = JSON.parse(data);
            return result;
        }
        catch (e) {
            return data;
        }
    };
    return Dialog;
}());
exports.Dialog = Dialog;
