// @microsoft/office-js-helpers/dist/helpers/storage.js
// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var dictionary_1 = require("./dictionary");
var utilities_1 = require("./utilities");
var StorageType;
(function (StorageType) {
    StorageType[StorageType["LocalStorage"] = 0] = "LocalStorage";
    StorageType[StorageType["SessionStorage"] = 1] = "SessionStorage";
})(StorageType = exports.StorageType || (exports.StorageType = {}));
/**
 * Helper for creating and querying Local Storage or Session Storage.
 * @see Uses {@link Dictionary} to create an in-memory copy of
 * the storage for faster reads. Writes update the actual storage.
 */
var Storage = (function (_super) {
    __extends(Storage, _super);
    /**
     * @constructor
     * @param {string} container Container name to be created in the LocalStorage.
     * @param {StorageType} type[optional] Storage Type to be used, defaults to Local Storage.
    */
    function Storage(container, type) {
        var _this = _super.call(this) || this;
        _this.container = container;
        _this._storage = null;
        _this._notifyObservers = function (event) { return Storage._observers.forEach(function (observer) { return observer(event); }); };
        type = type || StorageType.LocalStorage;
        _this.switchStorage(type);
        _this._registerStorageEvent();
        return _this;
    }
    /**
     * Switch the storage type.
     * Switches the storage type and then reloads the in-memory collection.
     *
     * @type {StorageType} type The desired storage to be used.
     */
    Storage.prototype.switchStorage = function (type) {
        this._storage = type === StorageType.LocalStorage ? localStorage : sessionStorage;
        if (!this._storage.hasOwnProperty(this.container)) {
            this._storage[this.container] = null;
        }
        this.load();
    };
    /**
     * Add an item.
     * Extends Dictionary's implementation of add, with a save to the storage.
     */
    Storage.prototype.add = function (item, value) {
        this.load();
        _super.prototype.add.call(this, item, value);
        this.save();
        return value;
    };
    /**
     * Add or Update an item.
     * Extends Dictionary's implementation of insert, with a save to the storage.
     */
    Storage.prototype.insert = function (item, value) {
        this.load();
        _super.prototype.insert.call(this, item, value);
        this.save();
        return value;
    };
    /**
     * Remove an item.
     * Extends Dictionary's implementation with a save to the storage.
     */
    Storage.prototype.remove = function (item) {
        this.load();
        var value = _super.prototype.remove.call(this, item);
        this.save();
        return value;
    };
    /**
     * Clear the storage.
     * Extends Dictionary's implementation with a save to the storage.
     */
    Storage.prototype.clear = function () {
        _super.prototype.clear.call(this);
        this._storage.removeItem(this.container);
    };
    /**
     * Clear all storages.
     * Completely clears both the localStorage and sessionStorage.
     */
    Storage.clearAll = function () {
        window.localStorage.clear();
        window.sessionStorage.clear();
    };
    /**
     * Saves the current state to the storage.
     */
    Storage.prototype.save = function () {
        this._storage.setItem(this.container, JSON.stringify(this.items));
    };
    /**
     * Refreshes the storage with the current localStorage values.
     */
    Storage.prototype.load = function () {
        var items = JSON.parse(this._storage.getItem(this.container));
        this.items = utilities_1.Utilities.extend({}, this.items, items);
    };
    /**
     * Registers an event handler for the window.storage event and
     * triggers the observer when the storage event is fired.
     *
     * The window.storage event is registered only once.
     */
    Storage.prototype.onStorage = function (observer) {
        Storage._observers.push(observer);
    };
    Storage.prototype._registerStorageEvent = function () {
        var _this = this;
        this.onStorage(function (event) { return _this.load(); });
        if (Storage._storageEventRegistered) {
            return;
        }
        window.onstorage = function (event) { return _this._notifyObservers(event); };
        Storage._storageEventRegistered = true;
    };
    return Storage;
}(dictionary_1.Dictionary));
Storage._observers = [];
exports.Storage = Storage;
